src/compiler/checker.ts

```diff
@@ -6835,6 +7028,23 @@ module ts {
                     });
                 }
             }
+            else if (node.kind === SyntaxKind.Decorator) {
+                let decorator = <Decorator>node;
+                switch (decorator.parent.kind) {
+                    case SyntaxKind.ClassDeclaration:
+                        args = [decorator.expression];
+                        break;
+                    case SyntaxKind.PropertyDeclaration:
+                        args = [decorator.expression, decorator.expression];
+                        break;
+                    case SyntaxKind.MethodDeclaration:
+                    case SyntaxKind.GetAccessor:
+                    case SyntaxKind.SetAccessor:
+                    case SyntaxKind.Parameter:
+                        args = [decorator.expression, decorator.expression, decorator.expression];

```

Comment:
I'll look into this. It's a bit of a chicken/egg problem in that I don't know the signature yet when I am deriving the arguments. I will look into your previous suggestion about using undefined.